File: ios/Mirobody/Data/Chat/ChatStreamClient.swift
import Foundation

/// POSTs `/api/chat` and decodes the Server-Sent-Events response into a stream of
/// `ChatStreamEvent`s. Mirrors `data/chat/ChatStreamClient.kt`.
///
/// Where Android uses OkHttp's `EventSource`, here we use `URLSession.bytes(for:)`
/// and split on lines. Each SSE `data:` line is one JSON chunk (`RawSseChunk`).
/// The returned `AsyncStream` finishes on the `end` event, on failure (after
/// emitting `.error`), or when the consuming task is cancelled.
final class ChatStreamClient {
    private let api: ApiClient
    private let session: URLSession

    init(api: ApiClient) {
        self.api = api
        // A dedicated session with a long request timeout: an SSE response stays
        // open for the whole turn, so the 60s timeout on the JSON client would cut
        // it off mid-stream.
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 3600
        cfg.timeoutIntervalForResource = 86_400
        self.session = URLSession(configuration: cfg)
    }

    func stream(
        _ request: ChatStreamRequest,
        attachments: [ChatAttachment] = []
    ) -> AsyncStream<ChatStreamEvent> {
        AsyncStream { continuation in
            let task = Task {
                do {
                    // With attachments, post a multipart form so the files ride along;
                    // otherwise the lighter JSON body. The server reads the same field
                    // names from either form (src/chat/params.cpp).
                    let req: URLRequest
                    if attachments.isEmpty {
                        let body = try api.encode(request)
                        req = try api.makeURLRequest(
                            path: "/api/chat",
                            method: "POST",
                            jsonBody: body,
                            accept: "text/event-stream"
                        )
                    } else {
                        req = try Self.multipartRequest(api: api, request: request, attachments: attachments)
                    }
                    var streamingReq = req
                    streamingReq.setValue("no-cache", forHTTPHeaderField: "Cache-Control")

                    let (bytes, response) = try await session.bytes(for: streamingReq)
                    guard let http = response as? HTTPURLResponse else {
                        continuation.yield(.error(message: "stream failed"))
                        continuation.finish()
                        return
                    }
                    guard (200..<300).contains(http.statusCode) else {
                        continuation.yield(.error(message: "HTTP \(http.statusCode)"))
                        continuation.finish()
                        return
                    }

                    for try await line in bytes.lines {
                        guard line.hasPrefix("data:") else { continue }   // skip event:/id:/comments
                        let payload = line.dropFirst("data:".count)
                            .trimmingCharacters(in: .whitespaces)
                        if payload.isEmpty { continue }
                        let event = Self.parseChunk(payload)
                        continuation.yield(event)
                        if case .end = event { break }
                    }
                    continuation.finish()
                } catch is CancellationError {
                    continuation.finish()
                } catch {
                    continuation.yield(.error(message: error.localizedDescription))
                    continuation.finish()
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// Builds a multipart/form-data `/api/chat` request: text fields + `file` parts.
    /// Mirrors the web client's FormData and the Android multipart path.
    private static func multipartRequest(
        api: ApiClient,
        request: ChatStreamRequest,
        attachments: [ChatAttachment]
    ) throws -> URLRequest {
        let boundary = "Boundary-\(UUID().uuidString)"
        var req = try api.makeURLRequest(path: "/api/chat", method: "POST", accept: "text/event-stream")
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        func append(_ s: String) { body.append(Data(s.utf8)) }
        func field(_ name: String, _ value: String) {
            append("--\(boundary)\r\n")
            append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n")
            append(value)
            append("\r\n")
        }
        field("agent", request.agent)
        field("provider", request.provider)
        field("question", request.question)
        field("language", request.language)
        field("session_id", request.sessionId)
        if let subject = request.subject, !subject.isEmpty { field("subject", subject) }
        for att in attachments {
            append("--\(boundary)\r\n")
            append("Content-Disposition: form-data; name=\"file\"; filename=\"\(att.fileName)\"\r\n")
            append("Content-Type: \(att.mimeType)\r\n\r\n")
            body.append(att.data)
            append("\r\n")
        }
        append("--\(boundary)--\r\n")
        req.httpBody = body
        return req
    }

    static func parseChunk(_ data: String) -> ChatStreamEvent {
        guard let chunk = try? JSONDecoder().decode(RawSseChunk.self, from: Data(data.utf8)) else {
            return .unknown(rawType: "parse_error", rawData: data)
        }
        func tool(_ make: (String) -> ChatStreamEvent) -> ChatStreamEvent {
            guard let id = chunk.toolId, !id.isEmpty else { return .unknown(rawType: chunk.type, rawData: data) }
            return make(id)
        }
        switch chunk.type {
        case "id":
            return .id(replyId: chunk.replyId ?? "")
        case "thinking":
            return .thinking(delta: chunk.content?.asString ?? "")
        case "reply":
            return .reply(delta: chunk.content?.asString ?? "")
        case "queryTitle":
            return tool { .queryTitle(toolId: $0, title: chunk.content?.asString ?? "") }
        case "queryArguments":
            return tool { .queryArguments(toolId: $0, argumentsJson: chunk.content?.asString ?? "") }
        case "queryDetail":
            return tool { .queryDetail(toolId: $0, detailJson: chunk.content?.asString ?? "") }
        case "image":
            let url = chunk.content?.asString ?? ""
            return url.isEmpty ? .unknown(rawType: chunk.type, rawData: data) : .image(url: url)
        case "chart":
            // The option is a nested JSON object; re-stringify it for echarts.setOption().
            let option = chunk.chart?.asString ?? ""
            return option.isEmpty ? .unknown(rawType: chunk.type, rawData: data) : .chart(optionJson: option)
        case "conversation":
            let id = chunk.content?.asString.nonBlank ?? chunk.conversationId ?? ""
            return id.isEmpty ? .unknown(rawType: chunk.type, rawData: data) : .conversation(id: id)
        case "heartbeat":
            return .heartbeat
        case "error":
            return .error(message: chunk.content?.asString ?? "")
        case "end":
            return .end
        case "costStatistics":
            if let node = chunk.content, let stats = try? node.decoded(CostStatistics.self) {
                return .stats(stats)
            }
            return .unknown(rawType: chunk.type, rawData: data)
        default:
            return .unknown(rawType: chunk.type, rawData: data)
        }
    }
}

File: android/app/src/main/java/ai/thetahealth/mirobody/data/chat/ChatStreamClient.kt
package ai.thetahealth.mirobody.data.chat

import ai.thetahealth.mirobody.data.chat.dto.ChatAttachment
import ai.thetahealth.mirobody.data.chat.dto.ChatStreamRequest
import ai.thetahealth.mirobody.data.chat.dto.ChatStreamEvent
import ai.thetahealth.mirobody.data.chat.dto.CostStatistics
import ai.thetahealth.mirobody.data.chat.dto.RawSseChunk
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources

/**
 * POST /api/chat as Server-Sent Events. Each SSE `data:` line is a JSON chunk
 * — see [RawSseChunk] / [ChatStreamEvent].
 */
class ChatStreamClient(
    client: OkHttpClient,
    private val json: Json,
) {
    // SSE callers must not have a BODY-level HttpLoggingInterceptor in the chain: it
    // calls source.request(Long.MAX_VALUE) on the response, draining the entire body
    // before the call returns and stalling event delivery until the server closes the
    // connection. Strip any logging interceptors for the streaming client; non-stream
    // requests still log via the original shared client.
    private val streamingClient: OkHttpClient = client.newBuilder()
        .apply {
            val kept = interceptors().filterNot { it is HttpLoggingInterceptor }
            interceptors().clear()
            interceptors().addAll(kept)
        }
        .build()
    private val sseFactory = EventSources.createFactory(streamingClient)

    fun stream(
        request: ChatStreamRequest,
        attachments: List<ChatAttachment> = emptyList(),
    ): Flow<ChatStreamEvent> = callbackFlow {
        // With attachments, post a multipart form so the files ride along (the server
        // stores them and references them on the turn); otherwise the lighter JSON body.
        // The server reads the same field names from either form (src/chat/params.cpp).
        val body: RequestBody = if (attachments.isEmpty()) {
            json.encodeToString(ChatStreamRequest.serializer(), request)
                .toRequestBody("application/json".toMediaType())
        } else {
            MultipartBody.Builder().setType(MultipartBody.FORM).apply {
                addFormDataPart("agent", request.agent)
                addFormDataPart("provider", request.provider)
                addFormDataPart("question", request.question)
                addFormDataPart("language", request.language)
                addFormDataPart("session_id", request.sessionId)
                request.subject?.takeIf { it.isNotBlank() }?.let { addFormDataPart("subject", it) }
                for (att in attachments) {
                    addFormDataPart(
                        "file",
                        att.fileName,
                        att.bytes.toRequestBody(att.mimeType.toMediaTypeOrNull()),
                    )
                }
            }.build()
        }
        val httpReq = Request.Builder()
            .url("http://placeholder.invalid/api/chat")
            .post(body)
            .header("Accept", "text/event-stream")
            .header("Cache-Control", "no-cache")
            .build()

        val listener = object : EventSourceListener() {
            override fun onOpen(eventSource: EventSource, response: Response) = Unit

            override fun onEvent(
                eventSource: EventSource,
                id: String?,
                type: String?,
                data: String,
            ) {
                val event = parseChunk(data)
                trySend(event)
                if (event is ChatStreamEvent.End) {
                    eventSource.cancel()
                    close()
                }
            }

            override fun onClosed(eventSource: EventSource) {
                close()
            }

            override fun onFailure(
                eventSource: EventSource,
                t: Throwable?,
                response: Response?,
            ) {
                val msg = t?.message
                    ?: response?.let { "HTTP ${it.code}" }
                    ?: "stream failed"
                trySend(ChatStreamEvent.Error(msg))
                close()
            }
        }

        val source = sseFactory.newEventSource(httpReq, listener)
        awaitClose { source.cancel() }
    }

    private fun parseChunk(data: String): ChatStreamEvent = try {
        val chunk = json.decodeFromString(RawSseChunk.serializer(), data)
        when (chunk.type) {
            "id" -> ChatStreamEvent.Id(chunk.replyId.orEmpty())
            "thinking" -> ChatStreamEvent.Thinking(chunk.content.asString())
            "reply" -> ChatStreamEvent.Reply(chunk.content.asString())
            "queryTitle" -> chunk.toolId
                ?.takeIf { it.isNotBlank() }
                ?.let { ChatStreamEvent.QueryTitle(it, chunk.content.asString()) }
                ?: ChatStreamEvent.Unknown(chunk.type, data)
            "queryArguments" -> chunk.toolId
                ?.takeIf { it.isNotBlank() }
                ?.let { ChatStreamEvent.QueryArguments(it, chunk.content.asString()) }
                ?: ChatStreamEvent.Unknown(chunk.type, data)
            "queryDetail" -> chunk.toolId
                ?.takeIf { it.isNotBlank() }
                ?.let { ChatStreamEvent.QueryDetail(it, chunk.content.asString()) }
                ?: ChatStreamEvent.Unknown(chunk.type, data)
            "image" -> {
                val url = chunk.content.asString()
                if (url.isBlank()) ChatStreamEvent.Unknown(chunk.type, data)
                else ChatStreamEvent.Image(url)
            }
            "chart" -> {
                // The option is a nested JSON object; re-stringify it for echarts.setOption().
                val option = chunk.chart
                if (option == null || option is JsonNull) ChatStreamEvent.Unknown(chunk.type, data)
                else ChatStreamEvent.Chart(option.toString())
            }
            "conversation" -> {
                val id = chunk.content.asString().ifBlank { chunk.conversationId.orEmpty() }
                if (id.isBlank()) ChatStreamEvent.Unknown(chunk.type, data)
                else ChatStreamEvent.Conversation(id)
            }
            "heartbeat" -> ChatStreamEvent.Heartbeat
            "error" -> ChatStreamEvent.Error(chunk.content.asString())
            "end" -> ChatStreamEvent.End
            "costStatistics" -> {
                val node = chunk.cost
                if (node != null && node !is JsonNull) {
                    // If decode throws, the outer try/catch turns it into Unknown("parse_error", ...).
                    ChatStreamEvent.Stats(json.decodeFromJsonElement(CostStatistics.serializer(), node))
                } else {
                    ChatStreamEvent.Unknown(chunk.type, data)
                }
            }
            else -> ChatStreamEvent.Unknown(chunk.type, data)
        }
    } catch (t: Throwable) {
        ChatStreamEvent.Unknown(rawType = "parse_error", rawData = data)
    }

    /** Backend emits `content` as a string for reply/thinking/error; tolerate either form. */
    private fun JsonElement?.asString(): String = when (this) {
        null -> ""
        is JsonPrimitive -> contentOrNull.orEmpty()
        else -> toString()
    }
}